src/coprocessor/codec/chunk/column.rs

```diff
@@ -0,0 +1,425 @@
+                    Datum::U64(self.get_u64(idx)?)
+                } else {
+                    Datum::I64(self.get_i64(idx)?)
+                }
+            }
+            types::FLOAT | types::DOUBLE => Datum::F64(self.get_f64(idx)?),
+            types::DATE | types::DATETIME | types::TIMESTAMP => Datum::Time(self.get_time(idx)?),
+            types::DURATION => Datum::Dur(self.get_duration(idx)?),
+            types::NEW_DECIMAL => Datum::Dec(self.get_decimal(idx)?),
+            types::JSON => Datum::Json(self.get_json(idx)?),
+            types::ENUM | types::BIT | types::SET => {
+                return Err(box_err!(
+                    "get datum with {:?} is not supported yet.",
+                    tp.get_tp()
+                ))
+            }
+            types::VARCHAR
+            | types::VAR_STRING
+            | types::STRING
+            | types::BLOB
+            | types::TINY_BLOB
+            | types::MEDIUM_BLOB
+            | types::LONG_BLOB => Datum::Bytes(self.get_bytes(idx).to_vec()),
+            _ => unreachable!(),
+        };
+        Ok(d)
+    }
+
+    pub fn append_datum(&mut self, data: &Datum) -> Result<()> {
+        match data {
+            Datum::Null => self.append_null(),
+            Datum::I64(v) => self.append_i64(*v),
+            Datum::U64(v) => self.append_u64(*v),
+            Datum::F64(ref v) => self.append_f64(*v),
+            Datum::Bytes(ref v) => self.append_bytes(v),
+            Datum::Dec(ref v) => self.append_decimal(v),
+            Datum::Dur(ref v) => self.append_duration(v),
+            Datum::Time(ref v) => self.append_time(v),
+            Datum::Json(ref v) => self.append_json(v),
+            _ => Err(box_err!("unsupported datum {:?}", data)),
+        }
+    }
+
+    pub fn new_fixed_column(fixed_len: usize, init_cap: usize) -> Column {
+        Column {
+            fixed_len,
+            data: Vec::with_capacity(fixed_len * init_cap),
+            null_bitmap: Vec::with_capacity(init_cap / 8),
+            ..Default::default()
+        }
+    }
+
+    pub fn new_var_len_column(init_cap: usize) -> Column {
+        let mut offsets = Vec::with_capacity(init_cap + 1);
+        offsets.push(0);
+        Column {
+            var_offsets: offsets,
+            data: Vec::with_capacity(4 * init_cap),
+            null_bitmap: Vec::with_capacity(init_cap / 8),
+            ..Default::default()
+        }
+    }
+
+    #[inline]
+    fn is_fixed(&self) -> bool {
+        self.fixed_len > 0
+    }
+
+    pub fn reset(&mut self) {
+        self.length = 0;
+        self.null_cnt = 0;
+        self.null_bitmap.clear();
+        if !self.var_offsets.is_empty() {
+            // The first offset is always 0, it makes slicing the data easier, we need to keep it.
+            self.var_offsets.truncate(1);
+        }
+        self.data.clear();
+    }
+
+    pub fn is_null(&self, row_idx: usize) -> bool {
+        if let Some(null_byte) = self.null_bitmap.get(row_idx >> 3) {
```

Comment:
/ 8?

Reply:
I think is more clear according to the document https://arrow.apache.org/docs/memory_layout.html